rust/main/agents/relayer/src/server/message_retry.rs

```diff
@@ -1,32 +1,87 @@
-use crate::settings::matching_list::MatchingList;
+use crate::{server::ENDPOINT_MESSAGES_QUEUE_SIZE, settings::matching_list::MatchingList};
+
 use axum::{extract::State, routing, Json, Router};
+
 use derive_new::new;
-use tokio::sync::broadcast::Sender;
+use serde::{Deserialize, Serialize};
+use tokio::sync::{broadcast::Sender, mpsc};
 
 const MESSAGE_RETRY_API_BASE: &str = "/message_retry";
 
-#[derive(new, Clone)]
+#[derive(new)]
 pub struct MessageRetryApi {
-    tx: Sender<MatchingList>,
+    retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryApiState {
+    pub retry_request_transmitter: Sender<MessageRetryRequest>,
+}
+
+#[derive(Clone, Debug)]
+pub struct MessageRetryRequest {
+    pub uuid: String,
+    pub pattern: MatchingList,
+    pub transmitter: mpsc::Sender<MessageRetryResponse>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MessageRetryResponse {
+    /// ID of the retry request
+    pub uuid: String,
+    /// how many pending operations were evaluated
+    pub evaluated: usize,
+    /// how many of the pending operations matched the retry request pattern
+    pub matched: u64,
 }
 
 async fn retry_message(
-    State(tx): State<Sender<MatchingList>>,
+    State(state): State<MessageRetryApiState>,
     Json(retry_req_payload): Json<MatchingList>,
-) -> String {
-    match tx.send(retry_req_payload) {
-        Ok(_) => "Moved message(s) to the front of the queue".to_string(),
-        // Technically it's bad practice to print the error message to the user, but
-        // this endpoint is for debugging purposes only.
-        Err(err) => format!("Failed to send retry request to the queue: {}", err),
+) -> Result<Json<MessageRetryResponse>, String> {
+    let uuid = uuid::Uuid::new_v4();
+    let uuid_string = uuid.to_string();
+
+    tracing::debug!(uuid = uuid_string, "Sending message retry request");
+
+    let (transmitter, mut receiver) = mpsc::channel(ENDPOINT_MESSAGES_QUEUE_SIZE);
```

Comment:
note this should be as big as the total number of chains the relayer operates. `ENDPOINT_MESSAGES_QUEUE_SIZE` is 100 which is already less than we have in prod I think. You can pass in the length of the configured `relayChains` when the server is constructed, and use that length as the channel size

Reply:
parsed here: https://github.com/kamiyaa/hyperlane-monorepo/blob/491f98110ed68cf3bd6838049ab3451fee775205/rust/main/agents/relayer/src/settings/mod.rs#L108

Reply:
actually much easier, pass the length as the max between `self.origin_chains` and `self.destination_chains` https://github.com/kamiyaa/hyperlane-monorepo/blob/491f98110ed68cf3bd6838049ab3451fee775205/rust/main/agents/relayer/src/relayer.rs#L66

Reply:
Why is it as big as total number of chains?
I thought there were just 3 queues? Or is it 3 queues per chain?

Reply:
hm, actually this just sets the max number of messages in the channel, for some reason I was thinking about the max number of consumers. The max consumers is `num_chains * num_queues`, where `num_chains` is 100+ and `num_queues` is 3 as you point out. 

Actually since: 1) the tokio `mpsc` applies backpressure by making senders wait; and 2) we expect very low message volumes since it's only relayer operators manually triggering this; then we can set the size to be `1` here 

just trying to be a bit memory conscious since the relayer started using up increasingly more recently

Reply:
Okay, I can update it to 1 :+1: